### ISSUE TITLE
Windows - complete confusion on install, PermissionError on urlwatch --edit

### ISSUE THREAD

**Issue by joseortiz3** [thp/urlwatch#223]
Hello, trying to get this working based on directions provided. I don't think the directions are sufficient to get this working for most people.

I'm running Windows 10 Pro, 64-bit Anaconda 5.1.0 w/ Python 3.

In accord with the directions provided, I installed the dependencies requested using `pip install minidb ... (etc)`.

After this, there seem to be **no further directions** for installation.

I cloned `urlwatch`. At first, I tried running `urlwatch` inside the directory. This did nothing. Then I tried  `python urlwatch` inside the project directory. This produces the message:

```
    You need to create C:\Users\Joey\AppData\Local\urlwatch\urlwatch\urls.yaml in order to use urlwatch.
    Use "urlwatch --edit" to open the file with your editor.
```

Ok, so I then ran `python urlwatch --edit` as requested. This produces the message

```
    Migrating cache: C:\Users\Joey\AppData\Local\urlwatch\urlwatch\cache -> C:\Users\Joey\AppData\Local\urlwatch\urlwatch\Cache\cache.db

Removing: cache.db
Traceback (most recent call last):
  File "urlwatch", line 107, in <module>
    urlwatch = Urlwatch(command_config, config_storage, cache_storage, urls_storage)
  File "D:\Downloads\urlwatch\lib\urlwatch\main.py", line 66, in __init__
    self.urlwatch_config.migrate_cache(self)
  File "D:\Downloads\urlwatch\lib\urlwatch\migration.py", line 82, in migrate_cache
    atomic_rename(cache_dir, cache_dir + '.migrated')
  File "D:\Downloads\urlwatch\lib\urlwatch\util.py", line 91, in atomic_rename
    os.rename(old_filename, new_filename)
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\Joey\\AppData\\Local\\urlwatch\\urlwatch\\cache' -> 'C:\\Users\\Joey\\AppData\\Local\\urlwatch\\urlwatch\\cache.migrated'
```
This message occurs whether or not I run with administrative privileges.

I then tried running `python setup.py install`. The install ran smoothly, but did nothing to change the above behavior. I am able to run `import urlwatch` inside python. But not sure how to get any of the functionality presented on the front page working.

So, in total, I'm very confused by the "getting started" instructions. If I'm this confused, I'm sure the majority of people encountering this seemingly-promising project are also confused. Help?

----------------------------------------------

_Edit: This confusion is probably because `pip install urlwatch` is not listed anywhere on the front page._

So now on a fresh installation of Anaconda 3.6 5.1.0, I ran `pip install urlwatch`, great. I see it installed some things in `Anaconda3/Scripts/`. I see a file with no extension (probably because of linux) called `urlwatch`. I run `python urlwatch`. It says the same messages as above about `Use "urlwatch --edit" to open the file with your editor.`. When I run `python urlwatch --edit`, the same permission error as above occurs. If I navigate to `AppData/Local/urlwatch`, I see the directory `/urlwatch/` is read-only, which might be why the permission errors occur. I would expect anyone using Windows would run into the same `PermissionError` problem.

**Comment by thp** [comment]
Seems to be a permission issue with folders on your drive?

Try this (replacing the folders seen there with the folders you are experiencing issues with):
https://answers.microsoft.com/en-us/windows/forum/windows_10-files-winpc/appdata-folder-access-denied/c24d4083-5f56-4d25-942b-5f750482d2eb

You should be able to install `urlwatch` using `pip`, the same way you installed the dependencies (in fact, installing using pip should also install the dependencies for you).

**Comment by coffeecer** [comment]
Hi. 

Question from a dummy:
Is there a simple way to deploy development environment and project in Windows or Google App Engine? Docker Container or something like that? 

Thank you.

**Comment by cfbao** [comment]
I think I've found the cause of this error.

If I'm not mistaken, urlwatch 1 stores cache in the directory "<appdir>/cache", whereas urlwatch 2 stores cache in the database "<appdir>/Cache/cache.db".
In the migration function, there's a test for the old cache directory
https://github.com/thp/urlwatch/blob/80e6194a7cc8c109c0ad022545996d6ba8338f20/lib/urlwatch/migration.py#L74

The problem is, Windows filenames are case insensitive. So this test will always pass on Windows, even for a fresh new install of urlwatch 2, because the directory "<appdir>/Cache" has already been created for the new cache database.
Trying to rename this directory will result in permission error, because "<appdir>/Cache/cache.db" is opened in urlwatch.

**Comment by cfbao** [comment]
The simplest way to fix this is probably to disable migration on Windows altogether.

Given the lack of earlier reports on this issue, there probably weren't many (if any) long time Windows users who cared about migration.

**Comment by joseortiz3** [comment]
@cfbao Of course the number of Windows users is artificially low because of these problems in the first place

**Comment by thp** [comment]
Fix pushed. Thanks for @cfbao for the analysis. Is somebody able to test this fix on Windows with a Git master checkout?

**Comment by cfbao** [comment]
I just tested the fix and it works.

**Comment by Hitaro** [comment]
Same error for me on latest version installed via pip3:

**Windows10>** `py -3 urlwatch`

    Migrating cache: C:\Users\<username>\AppData\Local\urlwatch\urlwatch\cache -> C:\Users\<username>\AppData\Local\urlwatch\urlwatch\Cache\cache.db
    
    Traceback (most recent call last):
     File "c:\users\<username>\appdata\local\programs\python\python37\scripts\urlwatch", line 107, in <module>
    urlwatch = Urlwatch(command_config, config_storage, cache_storage, urls_storage)
    File "C:\Users\<username>\AppData\Local\Programs\Python\Python37\lib\site-packages\urlwatch\main.py", line 66, in __init__
    self.urlwatch_config.migrate_cache(self)
    File "C:\Users\<username>\AppData\Local\Programs\Python\Python37\lib\site-packages\urlwatch\migration.py", line 80, in migrate_cache
    urlwatcher.cache_storage.restore(old_cache_storage.backup())
    File "C:\Users\<username>\AppData\Local\Programs\Python\Python37\lib\site-packages\urlwatch\storage.py", line 385, in restore
    for guid, data, timestamp, tries, etag in entries:
    File "C:\Users\<username>\AppData\Local\Programs\Python\Python37\lib\site-packages\urlwatch\storage.py", line 381, in backup
    data, timestamp, tries, etag = self.load(None, guid)
    ValueError: not enough values to unpack (expected 4, got 3)

**Comment by cfbao** [comment]
The fix isn't in any released version yet. You may want to manually download and install the version in master branch.

**Comment by thp** [comment]
This is fixed in 2.16, released today :)
